In [8]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity="all" # cell 的多行输出

# 客户价值预测模型

客户价值预测就是指预测客户在未来一段时间内能带来多少利润，其利润可能来自信用卡的年费、取现手续费、分期手续费、境外交易手续费等

分析出客户价值后，在进行营销、电话接听、催收、产品咨询等各项业务时，就可以针对高价值客户提供区别于普通客户的服务，以进一步挖掘这些高价值客户的价值，并提高他们的忠诚度

In [3]:
import pandas as pd
df=pd.read_excel('./data/客户价值数据表.xlsx')
df.head()

客户价值  历史贷款金额  贷款次数  学历    月收入  性别
0  1150    6488     2   2   9567   1
1  1157    5194     4   2  10767   0
2  1163    7066     3   2   9317   0
3   983    3550     3   2  10517   0
4  1205    7847     3   3  11267   1

In [5]:
X=df[['历史贷款金额', '贷款次数', '学历', '月收入', '性别']]
y=df['客户价值']

In [9]:
from sklearn.linear_model import LinearRegression
regr=LinearRegression()
regr.fit(X,y)
regr.intercept_
regr.coef_

LinearRegression()

-208.42004079958474

array([5.71421731e-02, 9.61723492e+01, 1.13452022e+02, 5.61326459e-02,
       1.97874093e+00])

In [12]:
import statsmodels.api as sm
X_=sm.add_constant(X)
est=sm.OLS(y,X_).fit()
est.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   客户价值   R-squared:                       0.571
Model:                            OLS   Adj. R-squared:                  0.553
Method:                 Least Squares   F-statistic:                     32.44
Date:                Fri, 09 Sep 2022   Prob (F-statistic):           6.41e-21
Time:                        12:55:39   Log-Likelihood:                -843.50
No. Observations:                 128   AIC:                             1699.
Df Residuals:                     122   BIC:                             1716.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       -208.4200    163.810     -1.272      0.206    -532.699     115.859
历史贷款金额         0.0571      0.010      5.945      0.000       0.038       0.076
贷款次数          96.1723     25.962      3.704      0.000      44.778     147.567
学历           113.4520     37.909      2.993      0.003      38.406     188.498
月收入            0.0561      0.019      2.941      0.004       0.018       0.094
性别             1.9787     32.286      0.061      0.951     -61.934      65.891
==============================================================================
Omnibus:                        1.597   Durbin-Watson:                   2.155
Prob(Omnibus):                  0.450   Jarque-Bera (JB):                1.538
Skew:                           0.264   Prob(JB):                        0.464
Kurtosis:                       2.900   Cond. No.                     1.28e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.28e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

R-squared值为0.571，Adj.R-squared值为0.553，整体拟合效果不是特别好

P值，可以发现大部分特征变量的P值都较小，的确与目标变量（即“客户价值”）显著相关

而“性别”这一特征变量的P值达到了0.951，即与目标变量没有显著相关性，在之后的建模中可以舍去“性别”这一特征变量

# 线性模型优缺点

线性回归模型的优点是可解释性很强且建模速度较快，缺点是模型相对简单，而且主要针对线性数据，局限性较大。在商业实战中，线性回归模型的应用相对较少